src/config.rs

```diff
@@ -0,0 +1,153 @@
+pub mod storage;
+pub mod settings;
+
+use std::collections::HashMap;
+use serde_derive::Deserialize;
+use serde_json::Value;
+use wildmatch::WildMatch;
+use crate::config::settings::{Setting, SettingInfo};
+
+const SETTINGS_JSON: &str = include_str!("./config/settings.json");
+
+#[derive(Debug, Deserialize)]
+struct JsonEntry {
+    key: String,
+    #[serde(rename = "type")]
+    entry_type: String,
+    default: String,
+    description: String,
+}
+
+trait StorageAdapter {
+    /// Retrieves a setting from the storage
+    fn get_setting(&self, key: &str) -> Option<Setting>;
+    /// Stores a given setting to the storage
+    fn set_setting(&mut self, key: &str, value: Setting);
+    /// Retrieves all the settings in the storage
+    fn get_all_settings(&self) -> HashMap<String, Setting>;
+}
+
+/// Configuration store is the place where the gosub engine can find all configurable options
+pub struct ConfigStore {
+    /// A hashmap of all settings so we can search o(1) time
+    settings: HashMap<String, Setting>,
+    /// A hashmap of all setting descriptions
+    settings_info: HashMap<String, SettingInfo>,
+    /// Keys of all settings so we can iterate keys easily
+    setting_keys: Vec<String>,
+    /// The storage adapter used for persisting and loading keys
+    storage: Box<dyn StorageAdapter>
+}
+
+impl ConfigStore {
+    /// Creates a new store with the given storage adapter and preloads the store if needed
+    fn new(storage: Box<dyn StorageAdapter>, preload: bool) -> Self {
+        let mut store = ConfigStore {
+            settings: HashMap::new(),
+            settings_info: HashMap::new(),
+            setting_keys: Vec::new(),
+            storage
+        };
+        
+        // Populate the settings from the json file
+        store.populate_settings();
+
+        // preload the settings if requested
+        if preload {
+            let all_settings = store.storage.get_all_settings();
```

Comment:
We might want to preload all settings from a storage engine, or lazy load them when needed. It depends a bit on the latency of the storage engine (ie: remote configuration would take a long time)